Tools/Scripts/pgo-profile

```diff
@@ -0,0 +1,142 @@
+    if len(string) > max_length:
+        return string[:max_length - 1] + u"…"
+    return string.ljust(max_length)
+
+
+def shell(command):
+    return subprocess.check_output(command, shell=True, stderr=subprocess.STDOUT).strip()
+
+
+def shortened(name):
+    return name if len(name) < 200 else name[0:99] + u"…" + name[-98:]
+
+
+def assert_directory(d):
+    if os.path.isdir(d):
+        return d
+    else:
+        print("No directory at ", d)
+        exit()
+
+
+def assert_file(f):
+    if os.path.isfile(f):
+        return f
+    else:
+        print("No file at ", f)
+        exit()
+
+
+def summarize_parser(subparsers):
+    parser = subparsers.add_parser("summarize", help="Dumps function names in a given .profraw file, sorted in descending order by function count")
+    parser.add_argument("file", help="Path to the .profraw file")
+    parser.set_defaults(func=summarize)
+    return parser
+
+
+def summarize(args):
+    file = assert_file(args.file)
+
+    lines = shell("xcrun -sdk macosx llvm-profdata show --all-functions --value-cutoff=10 \"{}\" | c++filt -n".format(file)).split(b"\n")
+
+    counts_and_functions = []
+
+    for line_number in range(len(lines)):
+        line = lines[line_number].strip()
+        if line.startswith(b"Function count: "):
+            count = int(line.split()[-1])
+            symbol = lines[line_number - 3].strip()[:-1].decode("utf-8")
+            counts_and_functions.append((count, symbol))
+
+    counts_and_functions.sort(reverse=True)
+    for count, name in counts_and_functions:
+        print(pad(str(count), 15), shortened(name))
+
+
+def merge_parser(subparsers):
+    parser = subparsers.add_parser("merge", help="Merge a pile of *.profraw files into the *.profdata files we can build with.")
+    parser.add_argument("directory", help="Path to the directory containing the *.profraw files")
+    parser.set_defaults(func=merge)
+    return parser
+
+
+def merge(args):
+    path = assert_directory(args.directory)
+
+    for lib in ["JavaScriptCore", "WebCore", "WebKit"]:
+        print("Merging", lib)
+        print(shell("xcrun -sdk macosx.internal llvm-profdata merge \"{}\"/{}*.profraw -output=\"{}/{}.profdata\"".format(path, lib, path, lib)))
+
+
+def combine_parser(subparsers):
+    parser = subparsers.add_parser("combine", help="Combine directories containing *.profdata files from different platforms together.")
+    parser.add_argument("--speedometer", default=None, help="Path to the directory containing the *.profdata files from a speedometer run.")
+    parser.add_argument("--jetstream", default=None, help="Path to the directory containing the *.profdata files from a jetstream run.")
+    parser.add_argument("--output", help="Path to the directory where the output will be placed.")
+    parser.set_defaults(func=combine)
+    return parser
+
+
+def combine(args):
+    out = assert_directory(args.output)
+    args = vars(args)
+    groups = [("speedometer", 0.6), ("jetstream", 0.4)]
```

Comment:
Why even have this script know about our benchmarks? Can we just provide the weight when passing in the path to the PGO profile. Perhaps it can default to equal weighting for all profiles.

Reply:
I figure we probably want to document the settings that we use somewhere? Otherwise we could just invoke the tool directly.